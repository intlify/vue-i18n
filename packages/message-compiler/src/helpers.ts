export const enum HelperNameMap {
  LIST = 'list',
  NAMED = 'named',
  PLURAL = 'plural',
  LINKED = 'linked',
  MESSAGE = 'message',
  TYPE = 'type',
  INTERPOLATE = 'interpolate',
  NORMALIZE = 'normalize'
}
