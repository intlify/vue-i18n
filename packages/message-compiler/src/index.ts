export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './parser'
export * from './compiler'
