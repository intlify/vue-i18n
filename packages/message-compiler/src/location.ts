// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export const LocationStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 }
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition(
  line: number,
  column: number,
  offset: number
): Position {
  return { line, column, offset } as Position
}

export function createLocation(
  start: Position,
  end: Position,
  source?: string
): SourceLocation {
  const loc = { start, end } as SourceLocation
  if (source != null) {
    loc.source = source
  }
  return loc
}
