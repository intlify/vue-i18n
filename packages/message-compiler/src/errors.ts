import { format } from '@intlify/shared'
import { SourceLocation } from './location'

export type CompileDomain =
  | 'tokenizer'
  | 'parser'
  | 'generator'
  | 'transformer'
  | 'compiler'

export interface CompileError extends SyntaxError {
  code: number
  domain?: CompileDomain
  location?: SourceLocation
}

export interface CreateCompileErrorOptions {
  domain?: CompileDomain
  messages?: { [code: number]: string }
  args?: unknown[]
}

export const enum CompileErrorCodes {
  // tokenizer error codes
  EXPECTED_TOKEN,
  INVALID_TOKEN_IN_PLACEHOLDER,
  UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER,
  UNKNOWN_ESCAPE_SEQUENCE,
  INVALID_UNICODE_ESCAPE_SEQUENCE,
  UNBALANCED_CLOSING_BRACE,
  UNTERMINATED_CLOSING_BRACE,
  EMPTY_PLACEHOLDER,
  NOT_ALLOW_NEST_PLACEHOLDER,
  INVALID_LINKED_FORMAT,

  // parser error codes
  MUST_HAVE_MESSAGES_IN_PLURAL,
  UNEXPECTED_EMPTY_LINKED_MODIFIER,
  UNEXPECTED_EMPTY_LINKED_KEY,
  UNEXPECTED_LEXICAL_ANALYSIS,

  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  __EXTEND_POINT__
}

/** @internal */
export const errorMessages: { [code: number]: string } = {
  // tokenizer error messages
  [CompileErrorCodes.EXPECTED_TOKEN]: `Expected token: '{0}'`,
  [CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER]: `Invalid token in placeholder: '{0}'`,
  [CompileErrorCodes.UNTERMINATED_SINGLE_QUOTE_IN_PLACEHOLDER]: `Unterminated single quote in placeholder`,
  [CompileErrorCodes.UNKNOWN_ESCAPE_SEQUENCE]: `Unknown escape sequence: \\{0}`,
  [CompileErrorCodes.INVALID_UNICODE_ESCAPE_SEQUENCE]: `Invalid unicode escape sequence: {0}`,
  [CompileErrorCodes.UNBALANCED_CLOSING_BRACE]: `Unbalanced closing brace`,
  [CompileErrorCodes.UNTERMINATED_CLOSING_BRACE]: `Unterminated closing brace`,
  [CompileErrorCodes.EMPTY_PLACEHOLDER]: `Empty placeholder`,
  [CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER]: `Not allowed nest placeholder`,
  [CompileErrorCodes.INVALID_LINKED_FORMAT]: `Invalid linked format`,
  // parser error messages
  [CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL]: `Plural must have messages`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER]: `Unexpected empty linked modifier`,
  [CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY]: `Unexpected empty linked key`,
  [CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS]: `Unexpected lexical analysis in token: '{0}'`
}

export function createCompileError<T extends number>(
  code: T,
  loc: SourceLocation | null,
  options: CreateCompileErrorOptions = {}
): CompileError {
  const { domain, messages, args } = options
  const msg = __DEV__
    ? format((messages || errorMessages)[code] || '', ...(args || []))
    : code
  const error = new SyntaxError(String(msg)) as CompileError
  error.code = code
  if (loc) {
    error.location = loc
  }
  error.domain = domain
  return error
}

/** @internal */
export function defaultOnError(error: CompileError): never {
  throw error
}
