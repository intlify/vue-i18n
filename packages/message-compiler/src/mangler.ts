import { CompileErrorCodes, createCompileError } from './errors'
import { NodeTypes } from './nodes'

import type {
  LinkedKeyNode,
  LinkedModifierNode,
  LinkedNode,
  ListNode,
  LiteralNode,
  MessageNode,
  NamedNode,
  Node,
  PluralNode,
  ResourceNode,
  TextNode
} from './nodes'

export const ERROR_DOMAIN = 'minifier'

/* eslint-disable @typescript-eslint/no-explicit-any */

export function mangle(node: Node) {
  node.t = node.type

  switch (node.type) {
    case NodeTypes.Resource: {
      const resource = node as ResourceNode
      mangle(resource.body)
      resource.b = resource.body
      delete (resource as any).body
      break
    }
    case NodeTypes.Plural: {
      const plural = node as PluralNode
      const cases = plural.cases
      for (let i = 0; i < cases.length; i++) {
        mangle(cases[i])
      }
      plural.c = cases
      delete (plural as any).cases
      break
    }
    case NodeTypes.Message: {
      const message = node as MessageNode
      const items = message.items
      for (let i = 0; i < items.length; i++) {
        mangle(items[i])
      }
      message.i = items
      delete (message as any).items
      if (message.static) {
        message.s = message.static
        delete (message as any).static
      }
      break
    }
    case NodeTypes.Text:
    case NodeTypes.Literal:
    case NodeTypes.LinkedModifier:
    case NodeTypes.LinkedKey: {
      const valueNode = node as
        | TextNode
        | LiteralNode
        | LinkedKeyNode
        | LinkedModifierNode
      if (valueNode.value) {
        valueNode.v = valueNode.value
        delete (valueNode as any).value
      }
      break
    }
    case NodeTypes.Linked: {
      const linked = node as LinkedNode
      mangle(linked.key)
      linked.k = linked.key
      delete (linked as any).key
      if (linked.modifier) {
        mangle(linked.modifier)
        linked.m = linked.modifier
        delete (linked as any).modifier
      }
      break
    }
    case NodeTypes.List: {
      const list = node as ListNode
      list.i = list.index
      delete (list as any).index
      break
    }
    case NodeTypes.Named: {
      const named = node as NamedNode
      named.k = named.key
      delete (named as any).key
      break
    }
    default:
      if (__DEV__) {
        throw createCompileError(
          CompileErrorCodes.UNHANDLED_MINIFIER_NODE_TYPE,
          null,
          {
            domain: ERROR_DOMAIN,
            args: [node.type]
          }
        )
      }
  }

  delete (node as any).type
}

/* eslint-enable @typescript-eslint/no-explicit-any */
