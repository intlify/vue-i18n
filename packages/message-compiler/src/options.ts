import { CompileError } from './errors'

export type CompileErrorHandler = (error: CompileError) => void
export type CompileCacheKeyHandler = (source: string) => string

export interface TokenizeOptions {
  location?: boolean
  onError?: CompileErrorHandler
}

export interface ParserOptions {
  location?: boolean
  onError?: CompileErrorHandler
}

export interface TransformOptions {
  onError?: CompileErrorHandler
}

export interface CodeGenOptions {
  mode?: 'normal' | 'arrow' // default normal
  breakLineCode?: '\n' | ';' // default newline
  needIndent?: boolean // default true
  onError?: CompileErrorHandler
  // Generate source map?
  // - Default: false
  sourceMap?: boolean
  // Filename for source map generation.
  // - Default: `message.intl`
  filename?: string
}

export type CompileOptions = {
  warnHtmlMessage?: boolean
  onCacheKey?: CompileCacheKeyHandler
} & TransformOptions &
  CodeGenOptions &
  ParserOptions &
  TokenizeOptions
