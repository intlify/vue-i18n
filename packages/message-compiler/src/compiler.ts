import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
