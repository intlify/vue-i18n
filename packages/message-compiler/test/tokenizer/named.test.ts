import { format } from '@intlify/shared'
import { TokenizeOptions } from '../../src/options'
import {
  CompileErrorCodes,
  CompileError,
  errorMessages
} from '../../src/errors'
import {
  createTokenizer,
  TokenTypes,
  ERROR_DOMAIN,
  parse
} from '../../src/tokenizer'

test('basic', () => {
  const tokenizer = createTokenizer('hi {name} !')
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: 'hi ',
    loc: {
      start: { line: 1, column: 1, offset: 0 },
      end: { line: 1, column: 4, offset: 3 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 4, offset: 3 },
      end: { line: 1, column: 5, offset: 4 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'name',
    loc: {
      start: { line: 1, column: 5, offset: 4 },
      end: { line: 1, column: 9, offset: 8 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 9, offset: 8 },
      end: { line: 1, column: 10, offset: 9 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: ' !',
    loc: {
      start: { line: 1, column: 10, offset: 9 },
      end: { line: 1, column: 12, offset: 11 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.EOF,
    loc: {
      start: { line: 1, column: 12, offset: 11 },
      end: { line: 1, column: 12, offset: 11 }
    }
  })
})

test('multiple', () => {
  const tokenizer = createTokenizer('{last} {middle}　{first}')
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 1, offset: 0 },
      end: { line: 1, column: 2, offset: 1 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'last',
    loc: {
      start: { line: 1, column: 2, offset: 1 },
      end: { line: 1, column: 6, offset: 5 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 6, offset: 5 },
      end: { line: 1, column: 7, offset: 6 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: ' ',
    loc: {
      start: { line: 1, column: 7, offset: 6 },
      end: { line: 1, column: 8, offset: 7 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 8, offset: 7 },
      end: { line: 1, column: 9, offset: 8 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'middle',
    loc: {
      start: { line: 1, column: 9, offset: 8 },
      end: { line: 1, column: 15, offset: 14 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 15, offset: 14 },
      end: { line: 1, column: 16, offset: 15 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: '　',
    loc: {
      start: { line: 1, column: 16, offset: 15 },
      end: { line: 1, column: 17, offset: 16 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 17, offset: 16 },
      end: { line: 1, column: 18, offset: 17 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'first',
    loc: {
      start: { line: 1, column: 18, offset: 17 },
      end: { line: 1, column: 23, offset: 22 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 23, offset: 22 },
      end: { line: 1, column: 24, offset: 23 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.EOF,
    loc: {
      start: { line: 1, column: 24, offset: 23 },
      end: { line: 1, column: 24, offset: 23 }
    }
  })
})

test('space', () => {
  const tokenizer = createTokenizer('hi {  name } !')
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: 'hi ',
    loc: {
      start: { line: 1, column: 1, offset: 0 },
      end: { line: 1, column: 4, offset: 3 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 4, offset: 3 },
      end: { line: 1, column: 5, offset: 4 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'name',
    loc: {
      start: { line: 1, column: 7, offset: 6 },
      end: { line: 1, column: 11, offset: 10 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 12, offset: 11 },
      end: { line: 1, column: 13, offset: 12 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: ' !',
    loc: {
      start: { line: 1, column: 13, offset: 12 },
      end: { line: 1, column: 15, offset: 14 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.EOF,
    loc: {
      start: { line: 1, column: 15, offset: 14 },
      end: { line: 1, column: 15, offset: 14 }
    }
  })
})

test('new line', () => {
  const tokenizer = createTokenizer('{last}\n{middle}\r\n{first}')
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 1, offset: 0 },
      end: { line: 1, column: 2, offset: 1 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'last',
    loc: {
      start: { line: 1, column: 2, offset: 1 },
      end: { line: 1, column: 6, offset: 5 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 6, offset: 5 },
      end: { line: 1, column: 7, offset: 6 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: '\n',
    loc: {
      start: { line: 1, column: 7, offset: 6 },
      end: { line: 2, column: 1, offset: 7 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 2, column: 1, offset: 7 },
      end: { line: 2, column: 2, offset: 8 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'middle',
    loc: {
      start: { line: 2, column: 2, offset: 8 },
      end: { line: 2, column: 8, offset: 14 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 2, column: 8, offset: 14 },
      end: { line: 2, column: 9, offset: 15 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: '\n',
    loc: {
      start: { line: 2, column: 9, offset: 15 },
      end: { line: 3, column: 1, offset: 17 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 3, column: 1, offset: 17 },
      end: { line: 3, column: 2, offset: 18 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: 'first',
    loc: {
      start: { line: 3, column: 2, offset: 18 },
      end: { line: 3, column: 7, offset: 23 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 3, column: 7, offset: 23 },
      end: { line: 3, column: 8, offset: 24 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.EOF,
    loc: {
      start: { line: 3, column: 8, offset: 24 },
      end: { line: 3, column: 8, offset: 24 }
    }
  })
})

describe('modulo cases', () => {
  test('basic named: hi %{name} !', () => {
    const tokenizer = createTokenizer('hi %{name} !')
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Text,
      value: 'hi ',
      loc: {
        start: { line: 1, column: 1, offset: 0 },
        end: { line: 1, column: 4, offset: 3 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Modulo,
      value: '%',
      loc: {
        start: { line: 1, column: 4, offset: 3 },
        end: { line: 1, column: 5, offset: 4 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.BraceLeft,
      value: '{',
      loc: {
        start: { line: 1, column: 5, offset: 4 },
        end: { line: 1, column: 6, offset: 5 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Named,
      value: 'name',
      loc: {
        start: { line: 1, column: 6, offset: 5 },
        end: { line: 1, column: 10, offset: 9 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.BraceRight,
      value: '}',
      loc: {
        start: { line: 1, column: 10, offset: 9 },
        end: { line: 1, column: 11, offset: 10 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Text,
      value: ' !',
      loc: {
        start: { line: 1, column: 11, offset: 10 },
        end: { line: 1, column: 13, offset: 12 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.EOF,
      loc: {
        start: { line: 1, column: 13, offset: 12 },
        end: { line: 1, column: 13, offset: 12 }
      }
    })
  })

  test('not modulo named: hi % {name} !', () => {
    const tokenizer = createTokenizer('hi % {name} !')
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Text,
      value: 'hi % ',
      loc: {
        start: { line: 1, column: 1, offset: 0 },
        end: { line: 1, column: 6, offset: 5 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.BraceLeft,
      value: '{',
      loc: {
        start: { line: 1, column: 6, offset: 5 },
        end: { line: 1, column: 7, offset: 6 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Named,
      value: 'name',
      loc: {
        start: { line: 1, column: 7, offset: 6 },
        end: { line: 1, column: 11, offset: 10 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.BraceRight,
      value: '}',
      loc: {
        start: { line: 1, column: 11, offset: 10 },
        end: { line: 1, column: 12, offset: 11 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Text,
      value: ' !',
      loc: {
        start: { line: 1, column: 12, offset: 11 },
        end: { line: 1, column: 14, offset: 13 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.EOF,
      loc: {
        start: { line: 1, column: 14, offset: 13 },
        end: { line: 1, column: 14, offset: 13 }
      }
    })
  })

  test('other placeholder syntax: hi %s !', () => {
    const tokenizer = createTokenizer('hi %s !')
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.Text,
      value: 'hi %s !',
      loc: {
        start: { line: 1, column: 1, offset: 0 },
        end: { line: 1, column: 8, offset: 7 }
      }
    })
    expect(tokenizer.nextToken()).toEqual({
      type: TokenTypes.EOF,
      loc: {
        start: { line: 1, column: 8, offset: 7 },
        end: { line: 1, column: 8, offset: 7 }
      }
    })
  })
})

test('underscore started', () => {
  const tokenizer = createTokenizer(
    `{_field} with the same value already exists.`
  )
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceLeft,
    value: '{',
    loc: {
      start: { line: 1, column: 1, offset: 0 },
      end: { line: 1, column: 2, offset: 1 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Named,
    value: '_field',
    loc: {
      start: { line: 1, column: 2, offset: 1 },
      end: { line: 1, column: 8, offset: 7 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.BraceRight,
    value: '}',
    loc: {
      start: { line: 1, column: 8, offset: 7 },
      end: { line: 1, column: 9, offset: 8 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.Text,
    value: ' with the same value already exists.',
    loc: {
      start: { line: 1, column: 9, offset: 8 },
      end: { line: 1, column: 45, offset: 44 }
    }
  })
  expect(tokenizer.nextToken()).toEqual({
    type: TokenTypes.EOF,
    loc: {
      start: { line: 1, column: 45, offset: 44 },
      end: { line: 1, column: 45, offset: 44 }
    }
  })
})

describe('errors', () => {
  let errors: CompileError[], options: TokenizeOptions
  beforeEach(() => {
    errors = []
    options = {
      onError: err => {
        errors.push({ ...err, message: err.message })
      }
    } as TokenizeOptions
  })

  test('empty placeholder', () => {
    parse(`hi, {} !`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.EMPTY_PLACEHOLDER,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.EMPTY_PLACEHOLDER],
        location: {
          start: {
            line: 1,
            offset: 5,
            column: 6
          },
          end: {
            line: 1,
            offset: 5,
            column: 6
          }
        }
      }
    ] as CompileError[])
  })

  test('nest placeholder', () => {
    parse(`hi, {{ kazupon`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.NOT_ALLOW_NEST_PLACEHOLDER],
        location: {
          start: {
            line: 1,
            offset: 5,
            column: 6
          },
          end: {
            line: 1,
            offset: 5,
            column: 6
          }
        }
      },
      {
        code: CompileErrorCodes.UNTERMINATED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNTERMINATED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 7,
            column: 8
          },
          end: {
            line: 1,
            offset: 14,
            column: 15
          }
        }
      }
    ] as CompileError[])
  })

  test('close brace only', () => {
    parse(`hi, :-} !`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.UNBALANCED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNBALANCED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 6,
            column: 7
          },
          end: {
            line: 1,
            offset: 6,
            column: 7
          }
        }
      }
    ] as CompileError[])
  })

  test('not close brace at sentence', () => {
    parse(`hi { name !`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.UNTERMINATED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNTERMINATED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 10,
            column: 11
          },
          end: {
            line: 1,
            offset: 10,
            column: 11
          }
        }
      }
    ] as CompileError[])
  })

  test('not close brace at EOF', () => {
    parse(`hi {name`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.UNTERMINATED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNTERMINATED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 4,
            column: 5
          },
          end: {
            line: 1,
            offset: 8,
            column: 9
          }
        }
      }
    ] as CompileError[])
  })

  test('not close brace in plural message', () => {
    parse(`hi {  | hello {name} !`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.UNTERMINATED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNTERMINATED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 6,
            column: 7
          },
          end: {
            line: 1,
            offset: 6,
            column: 7
          }
        }
      }
    ] as CompileError[])
  })

  test('not close brace with linked', () => {
    parse(`hi { @:name !`, options)
    expect(errors).toEqual([
      {
        code: CompileErrorCodes.UNTERMINATED_CLOSING_BRACE,
        domain: ERROR_DOMAIN,
        message: errorMessages[CompileErrorCodes.UNTERMINATED_CLOSING_BRACE],
        location: {
          start: {
            line: 1,
            offset: 5,
            column: 6
          },
          end: {
            line: 1,
            offset: 5,
            column: 6
          }
        }
      }
    ] as CompileError[])
  })
  ;[`$`, `-`].forEach(ch => {
    test(`invalid '${ch}' in placeholder`, () => {
      parse(`hi {${ch}} !`, options)
      expect(errors).toEqual([
        {
          code: CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER,
          domain: ERROR_DOMAIN,
          message: format(
            errorMessages[CompileErrorCodes.INVALID_TOKEN_IN_PLACEHOLDER],
            ch
          ),
          location: {
            start: {
              line: 1,
              offset: 4,
              column: 5
            },
            end: {
              line: 1,
              offset: 5,
              column: 6
            }
          }
        }
      ] as CompileError[])
    })
  })
})
