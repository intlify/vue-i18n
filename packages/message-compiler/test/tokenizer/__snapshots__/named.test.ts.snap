// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modulo cases multiple  %{action} issue %s %{code} ! 1`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": 0,
  "value": " ",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 2`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
    "start": Object {
      "column": 2,
      "line": 1,
      "offset": 1,
    },
  },
  "type": 4,
  "value": "%",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 3`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
    "start": Object {
      "column": 3,
      "line": 1,
      "offset": 2,
    },
  },
  "type": 2,
  "value": "{",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 4`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 10,
      "line": 1,
      "offset": 9,
    },
    "start": Object {
      "column": 4,
      "line": 1,
      "offset": 3,
    },
  },
  "type": 5,
  "value": "action",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 5`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
    "start": Object {
      "column": 10,
      "line": 1,
      "offset": 9,
    },
  },
  "type": 3,
  "value": "}",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 6`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 21,
      "line": 1,
      "offset": 20,
    },
    "start": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
  },
  "type": 0,
  "value": " issue %s ",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 7`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": Object {
      "column": 21,
      "line": 1,
      "offset": 20,
    },
  },
  "type": 4,
  "value": "%",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 8`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "start": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
  },
  "type": 2,
  "value": "{",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 9`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
    "start": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
  },
  "type": 5,
  "value": "code",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 10`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 28,
      "line": 1,
      "offset": 27,
    },
    "start": Object {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
  },
  "type": 3,
  "value": "}",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 11`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 30,
      "line": 1,
      "offset": 29,
    },
    "start": Object {
      "column": 28,
      "line": 1,
      "offset": 27,
    },
  },
  "type": 0,
  "value": " !",
}
`;

exports[`modulo cases multiple  %{action} issue %s %{code} ! 12`] = `
Object {
  "loc": Object {
    "end": Object {
      "column": 30,
      "line": 1,
      "offset": 29,
    },
    "start": Object {
      "column": 30,
      "line": 1,
      "offset": 29,
    },
  },
  "type": 14,
}
`;
